src/amber/controller/filters.cr

```diff
@@ -0,0 +1,83 @@
+module Amber::Controller
+  module Callbacks
+    macro included
+      include Amber::DSL::Callbacks
+      protected property filters : Filters = Filters.new
+
+      protected def run_before_filter(action)
+        if self.responds_to? :before_filters
+          self.before_filters
+          @filters.run(:before, action)
+        end
+      end
+
+      protected def run_after_filter(action)
+        if self.responds_to? :after_filters
+          self.after_filters
+          @filters.run(:after, action)
+        end
+      end
+    end
+  end
+
+  record Filter, precedence : Symbol, action : Symbol, blk : -> Nil do
+  end
+
+  # Builds a BeforeAction filter.
+  #
+  # The yielded object has an `only` method that accepts two arguments,
+  # a key (`Symbol`) and a block ->`Nil`.
+  #
+  # ```
+  # FilterChainBuilder.build do |b|
+  #   filter :index, :show { some_method }
+  #   filter :delete { }
+  # end
+  # ```
+  record FilterBuilder, filters : Filters, precedence : Symbol do
+    def only(action : Symbol, &block : -> Nil)
+      add(action, &block)
+    end
+
+    def only(actions : Array(Symbol), &block : -> Nil)
+      actions.each { |action| add(action, &block) }
+    end
+
+    def add(action, &block : -> Nil)
+      filters.add Filter.new(precedence, action, block)
+    end
+  end
+
+  class Filters
+    include Enumerable({Symbol, Array(Filter)})
+
+    property filters = {} of Symbol => Array(Filter)
+
+    def register(precedence : Symbol)
```

Comment:
would specifying nil return type for this method allow us to remove the explicit `nil` return when calling this method [below](https://github.com/Amber-Crystal/amber/pull/30/files#diff-38ccdec39f6cb0c091906082b1f27cddR7), i.e. `def register(precendence: Symbol) : Nil`